File: rust/vectorpin/src/signer.rs
// Copyright 2025 Jascha Wanger / Tarnover, LLC
// SPDX-License-Identifier: Apache-2.0

//! Pin signing.
//!
//! Wraps an Ed25519 signing key plus a `kid` (key id) so verifiers can
//! route signatures during key rotation. Use [`Signer::generate`] for
//! tests and demos; load production keys from a managed secret store
//! via [`Signer::from_private_bytes`].
//!
//! # Examples
//!
//! ```
//! use vectorpin::Signer;
//!
//! let signer = Signer::generate("prod-2026-05".to_string()).unwrap();
//! let v: Vec<f32> = vec![0.1, 0.2, 0.3];
//! let pin = signer.pin("hello", "text-embedding-3-large", v.as_slice()).unwrap();
//! assert_eq!(pin.kid, "prod-2026-05");
//! assert_eq!(pin.sig.len(), 64); // Ed25519 signature
//! ```
//!
//! For deterministic signing (test fixtures, reproducible CI builds),
//! use [`PinOptions`] to supply an explicit timestamp and dtype:
//!
//! ```
//! use vectorpin::signer::{PinOptions, Signer};
//! use vectorpin::VecDtype;
//!
//! let signer = Signer::generate("test".to_string()).unwrap();
//! let v: Vec<f32> = vec![0.1, 0.2, 0.3];
//! let opts = PinOptions {
//!     dtype: Some(VecDtype::F32),
//!     timestamp: Some("2026-05-05T12:00:00Z".to_string()),
//!     ..PinOptions::default()
//! };
//! let pin = signer
//!     .pin_with_options("hello", "test-model", v.as_slice(), opts)
//!     .unwrap();
//! assert_eq!(pin.header.ts, "2026-05-05T12:00:00Z");
//! ```

use std::collections::BTreeMap;

use ed25519_dalek::{Signer as _, SigningKey, VerifyingKey};
use zeroize::Zeroizing;

use crate::attestation::{Pin, PinHeader, PROTOCOL_VERSION};
use crate::hash::{hash_text, hash_vector, VecDtype, VectorRef};

/// Errors raised by signer construction or pinning.
#[derive(Debug, thiserror::Error)]
pub enum SignerError {
    /// `key_id` was empty.
    #[error("key_id must be non-empty")]
    EmptyKeyId,
    /// Private key bytes were the wrong length (must be 32).
    #[error("private key must be exactly 32 bytes, got {0}")]
    BadKeyLength(usize),
    /// Vector was empty or otherwise unrepresentable.
    #[error("invalid vector: {0}")]
    InvalidVector(&'static str),
}

/// Produces signed [`Pin`] attestations.
///
/// A `Signer` holds one Ed25519 private key plus a stable identifier
/// (`kid`) that gets embedded in every pin it produces. Verifiers use
/// the `kid` to look up the matching public key in their registry, so
/// rotating signing keys is a matter of issuing a new `(kid, key)` pair
/// and accepting both the old and new `kid` values during the rotation
/// window — no protocol changes required.
///
/// # Secret material
///
/// [`Signer::generate`] is for tests, demos, and one-off CLI tools.
/// In production, hold the 32-byte private seed in a managed secrets
/// store (HSM, KMS, sealed env var) and instantiate via
/// [`Signer::from_private_bytes`]. [`Signer::private_key_bytes`] is
/// provided for backup/key-export workflows; treat its output as
/// secret.
pub struct Signer {
    signing_key: SigningKey,
    key_id: String,
}

impl Signer {
    /// Generate a fresh Ed25519 signer. Tests and demos only.
    ///
    /// Returns [`SignerError::EmptyKeyId`] if `key_id` is empty so the
    /// constructor matches the contract of [`Signer::from_private_bytes`].
    pub fn generate(key_id: String) -> Result<Self, SignerError> {
        if key_id.is_empty() {
            return Err(SignerError::EmptyKeyId);
        }
        let mut rng = rand::rngs::OsRng;
        Ok(Signer {
            signing_key: SigningKey::generate(&mut rng),
            key_id,
        })
    }

    /// Load a signer from a 32-byte raw Ed25519 private seed.
    pub fn from_private_bytes(raw: &[u8], key_id: String) -> Result<Self, SignerError> {
        if key_id.is_empty() {
            return Err(SignerError::EmptyKeyId);
        }
        let bytes: [u8; 32] = raw
            .try_into()
            .map_err(|_| SignerError::BadKeyLength(raw.len()))?;
        Ok(Signer {
            signing_key: SigningKey::from_bytes(&bytes),
            key_id,
        })
    }

    /// Identifier of the key used to sign — published in produced pins as `kid`.
    pub fn key_id(&self) -> &str {
        &self.key_id
    }

    /// 32-byte raw Ed25519 public key. This is what verifiers register.
    pub fn public_key_bytes(&self) -> [u8; 32] {
        VerifyingKey::from(&self.signing_key).to_bytes()
    }

    /// 32-byte raw Ed25519 private seed, wrapped in [`Zeroizing`] so the
    /// buffer is wiped from memory on drop. Treat the contents as
    /// secret; deref the returned value to access the raw `[u8; 32]`.
    pub fn private_key_bytes(&self) -> Zeroizing<[u8; 32]> {
        Zeroizing::new(self.signing_key.to_bytes())
    }

    /// Create a [`Pin`] for `(source, model, vector)`.
    ///
    /// `vector` accepts anything that converts into a [`VectorRef`],
    /// which includes `&[f32]` and `&[f64]`. The dtype written into
    /// the header defaults to the native dtype of the slice.
    pub fn pin<'a>(
        &self,
        source: &str,
        model: &str,
        vector: impl Into<VectorRef<'a>>,
    ) -> Result<Pin, SignerError> {
        self.pin_with_options(source, model, vector, PinOptions::default())
    }

    /// [`Self::pin`] with explicit dtype, model hash, timestamp, and
    /// `extra` map. Used by deterministic test-vector generation.
    pub fn pin_with_options<'a>(
        &self,
        source: &str,
        model: &str,
        vector: impl Into<VectorRef<'a>>,
        opts: PinOptions,
    ) -> Result<Pin, SignerError> {
        let vector = vector.into();
        if vector.is_empty() {
            return Err(SignerError::InvalidVector("empty vector"));
        }

        let dtype = opts.dtype.unwrap_or_else(|| vector.native_dtype());
        let ts = opts.timestamp.unwrap_or_else(now_utc_iso8601);

        let vec_dim = u32::try_from(vector.len())
            .map_err(|_| SignerError::InvalidVector("vec_dim exceeds u32"))?;

        let header = PinHeader {
            v: PROTOCOL_VERSION,
            model: model.to_owned(),
            model_hash: opts.model_hash,
            source_hash: hash_text(source),
            vec_hash: hash_vector(vector, dtype),
            vec_dtype: dtype.as_str().to_owned(),
            vec_dim,
            ts,
            extra: opts.extra,
        };

        let signature = self.signing_key.sign(&header.canonicalize());
        Ok(Pin {
            header,
            kid: self.key_id.clone(),
            sig: signature.to_bytes().to_vec(),
        })
    }
}

/// Optional knobs for [`Signer::pin_with_options`].
///
/// Defaults match what `Signer::pin` uses: native dtype, no model hash,
/// current UTC time, no extra metadata.
#[derive(Debug, Default, Clone)]
pub struct PinOptions {
    /// Override the canonical dtype the vector is hashed under.
    pub dtype: Option<VecDtype>,
    /// Optional content hash of the model weights.
    pub model_hash: Option<String>,
    /// Optional explicit timestamp string (must already be in
    /// `YYYY-MM-DDTHH:MM:SSZ` form).
    pub timestamp: Option<String>,
    /// Optional string-to-string `extra` map. Values are signed.
    pub extra: BTreeMap<String, String>,
}

fn now_utc_iso8601() -> String {
    // Produce a second-resolution UTC timestamp in `YYYY-MM-DDTHH:MM:SSZ`
    // form, matching the existing wire-format contract. The v1.1 branch
    // is responsible for any tightening of this format.
    let now = time::OffsetDateTime::now_utc();
    let fmt = time::macros::format_description!(
        "[year]-[month]-[day]T[hour]:[minute]:[second]Z"
    );
    now.format(&fmt)
        .expect("UTC OffsetDateTime always formats with a fixed description")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pin_round_trip_basic() {
        let signer = Signer::generate("test".into()).unwrap();
        let v: Vec<f32> = vec![1.0, 2.0, 3.0];
        let pin = signer.pin("hello", "model", v.as_slice()).unwrap();
        assert_eq!(pin.kid, "test");
        assert_eq!(pin.header.vec_dim, 3);
        assert_eq!(pin.header.vec_dtype, "f32");
        assert_eq!(pin.sig.len(), 64);
    }

    #[test]
    fn generate_rejects_empty_kid() {
        let res = Signer::generate("".into());
        assert!(matches!(res, Err(SignerError::EmptyKeyId)));
    }

    #[test]
    fn from_private_bytes_rejects_empty_kid() {
        let res = Signer::from_private_bytes(&[0u8; 32], "".into());
        assert!(matches!(res, Err(SignerError::EmptyKeyId)));
    }

    #[test]
    fn from_private_bytes_rejects_bad_length() {
        let res = Signer::from_private_bytes(&[0u8; 16], "k".into());
        assert!(matches!(res, Err(SignerError::BadKeyLength(16))));
    }

    #[test]
    fn private_seed_round_trip() {
        let signer = Signer::generate("k".into()).unwrap();
        let seed = signer.private_key_bytes();
        let restored = Signer::from_private_bytes(seed.as_ref(), "k".into()).unwrap();
        assert_eq!(signer.public_key_bytes(), restored.public_key_bytes());
    }
}

File: src/vectorpin/signer.py
# Copyright 2025 Jascha Wanger / Tarnover, LLC
# SPDX-License-Identifier: Apache-2.0
"""Pin signing.

We use Ed25519 because:
  - Signatures are 64 bytes (compact in DB metadata).
  - Public keys are 32 bytes.
  - Deterministic — same input always produces the same signature.
  - Widely supported across languages (matters for Symbiont's Rust runtime
    and any future MCP server implementations).

A Signer wraps a single (private_key, key_id) pair. Key rotation is a
deployment concern: issue a new (key_id, key) and have the verifier
accept multiple kids during the rotation window.
"""

from __future__ import annotations

import math
import unicodedata
from datetime import UTC, datetime

import numpy as np
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

from vectorpin.attestation import (
    PROTOCOL_VERSION,
    Pin,
    PinHeader,
    _check_nfc,
    _check_string_safe,
)
from vectorpin.hash import CanonicalDtype, hash_text, hash_vector


def _normalize_str(value: str, field_name: str) -> str:
    """NFC-normalize + reject control/bidi characters in a string input.

    Signers tolerate non-NFC input (they normalize before signing) but
    still reject structurally hostile characters so the signed pin is
    always parseable by a strict verifier.
    """
    nfc = unicodedata.normalize("NFC", value)
    _check_string_safe(nfc, field_name)
    return nfc


class Signer:
    """Produces Pin attestations for embeddings.

    A Signer holds one ed25519 private key. The corresponding public key
    is published with `key_id` so verifiers can route signatures to the
    right key during rotation.
    """

    def __init__(self, private_key: Ed25519PrivateKey, key_id: str):
        if not key_id:
            raise ValueError("key_id must be non-empty")
        # Normalize/validate the key id once at construction so every
        # subsequent Pin emits a header parseable by a strict verifier.
        normalized = _normalize_str(key_id, "key_id")
        self._private_key = private_key
        self._key_id = normalized

    @classmethod
    def generate(cls, key_id: str) -> Signer:
        """Generate a fresh ed25519 signer. Tests and demos only.

        Production deployments should load private keys from a managed
        secrets store, not generate them per-process.
        """
        return cls(Ed25519PrivateKey.generate(), key_id)

    @classmethod
    def from_private_bytes(cls, raw: bytes, key_id: str) -> Signer:
        """Load a signer from a 32-byte ed25519 private seed."""
        return cls(Ed25519PrivateKey.from_private_bytes(raw), key_id)

    @classmethod
    def from_pem(
        cls,
        pem: bytes,
        key_id: str,
        password: bytes | None = None,
        *,
        allow_unencrypted: bool = False,
    ) -> Signer:
        """Load a signer from PEM-encoded PKCS#8 ed25519 key material.

        Callers must either provide a `password` to decrypt an
        encrypted PEM, or set `allow_unencrypted=True` to opt in to
        loading an unencrypted file. The default is to refuse:
        unencrypted private keys on disk are a footgun, and we want a
        positive confirmation that the caller knew the file lacked
        encryption.
        """
        if password is None and not allow_unencrypted:
            raise ValueError(
                "PEM is unencrypted; pass allow_unencrypted=True to confirm"
            )
        key = serialization.load_pem_private_key(pem, password=password)
        if not isinstance(key, Ed25519PrivateKey):
            raise TypeError(f"expected Ed25519PrivateKey, got {type(key).__name__}")
        return cls(key, key_id)

    @property
    def key_id(self) -> str:
        return self._key_id

    def public_key(self) -> Ed25519PublicKey:
        return self._private_key.public_key()

    def public_key_bytes(self) -> bytes:
        """32-byte raw ed25519 public key — what verifiers actually need."""
        return self.public_key().public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw,
        )

    def private_key_bytes(self) -> bytes:
        """32-byte raw ed25519 private seed. Treat as a secret."""
        return self._private_key.private_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PrivateFormat.Raw,
            encryption_algorithm=serialization.NoEncryption(),
        )

    def pin(
        self,
        source: str,
        model: str,
        vector: np.ndarray,
        *,
        vec_dtype: CanonicalDtype = "f32",
        model_hash: str | None = None,
        timestamp: datetime | None = None,
        extra: dict[str, str] | None = None,
    ) -> Pin:
        """Create a Pin for (source, model, vector).

        Per §3.2, vectors containing NaN, +inf, or -inf are rejected at
        sign time. Per §3.1, every string-typed input is NFC-normalized
        and checked for control characters and bidi overrides.

        Args:
            source: The exact source text the embedding was produced from.
                Hashed and committed to; the verifier needs the same text
                to validate the pin.
            model: Embedding model identifier, e.g. 'text-embedding-3-large'.
                Treat as opaque — the verifier just compares strings.
            vector: 1-D numpy array, the embedding itself.
            vec_dtype: Canonical dtype to hash under. Default 'f32'.
            model_hash: Optional content hash of the model weights, if
                pinning to a specific local model file.
            timestamp: Optional explicit timestamp. Default: now (UTC).
            extra: Optional string-to-string metadata committed under the
                signature. Use sparingly — every key adds attack surface.

        Returns:
            A signed Pin. Serialize with `pin.to_json()` and store
            alongside the vector in the DB metadata.
        """
        # Reject NaN / Inf at sign time so a signer never commits to a
        # vector value with ambiguous hash semantics. The cast to the
        # canonical dtype happens here too so we catch overflows that
        # would silently become +inf in f32.
        target = np.dtype("<f4") if vec_dtype == "f32" else np.dtype("<f8")
        if vector.ndim != 1:
            raise ValueError(f"expected 1-D vector, got shape {vector.shape}")
        cast = vector.astype(target, copy=False)
        if not np.isfinite(cast).all():
            raise ValueError(
                "vector contains NaN or infinity; refusing to sign"
            )
        # Sanity-check: the unrounded array was finite too. Catches a
        # caller passing inf in f64 that got clipped by the cast.
        if not np.isfinite(vector).all() and not all(
            math.isfinite(float(x)) for x in vector
        ):
            raise ValueError(
                "vector contains NaN or infinity; refusing to sign"
            )

        # Normalize string inputs. The signer is intentionally
        # tolerant here: if a caller passes an NFD string we silently
        # NFC it, but we still reject control chars and bidi overrides.
        model_norm = _normalize_str(model, "model")
        source_norm = unicodedata.normalize("NFC", source)

        extra_norm: dict[str, str] = {}
        if extra:
            for k, val in extra.items():
                if not isinstance(k, str) or not isinstance(val, str):
                    raise ValueError(
                        "extra must be a map of str -> str"
                    )
                k_norm = _normalize_str(k, f"extra key {k!r}")
                v_norm = _normalize_str(val, f"extra[{k!r}]")
                extra_norm[k_norm] = v_norm

        if timestamp is None:
            timestamp = datetime.now(UTC)
        ts_iso = timestamp.astimezone(UTC).strftime("%Y-%m-%dT%H:%M:%SZ")
        # Defensive: the strftime above should always emit the v2 ts
        # format, but verify so a buggy platform locale can't sneak
        # something else through.
        _check_nfc(ts_iso, "ts")

        header = PinHeader(
            v=PROTOCOL_VERSION,
            kid=self._key_id,
            model=model_norm,
            model_hash=model_hash,
            source_hash=hash_text(source_norm),
            vec_hash=hash_vector(cast, vec_dtype),
            vec_dtype=vec_dtype,
            vec_dim=int(cast.shape[0]),
            ts=ts_iso,
            extra=extra_norm,
        )
        sig = self._private_key.sign(header.canonicalize())
        return Pin(header=header, sig=sig)
